contracts/tg4-stake/src/contract.rs

```diff
@@ -265,51 +330,70 @@ pub fn execute_slash<Q: CustomQuery>(
     let cfg = CONFIG.load(deps.storage)?;
     let addr = deps.api.addr_validate(&addr)?;
 
-    // update the addr's stake
+    let liquid_stake = STAKE.may_load(deps.storage, &addr)?;
+    let vesting_stake = STAKE_VESTING.may_load(deps.storage, &addr)?;
+
     // If address doesn't match anyone, leave early
-    let stake = match STAKE.may_load(deps.storage, &addr)? {
-        Some(s) => s,
-        None => return Ok(Response::new()),
-    };
-    let mut slashed = stake * portion;
-    let new_stake = STAKE.update(deps.storage, &addr, |stake| -> StdResult<_> {
-        Ok(stake.unwrap_or_default().checked_sub(slashed)?)
+    if liquid_stake.is_none() && vesting_stake.is_none() {
+        return Ok(Response::new());
+    }
+
+    // slash the liquid stake, if any
+    let mut liquid_slashed = liquid_stake.map(|s| s * portion).unwrap_or_default();
+    let new_liquid_stake = STAKE.update(deps.storage, &addr, |stake| -> StdResult<_> {
+        Ok(stake.unwrap_or_default().sub(liquid_slashed))
     })?;
 
-    // slash the claims
-    slashed += claims().slash_claims_for_addr(deps.storage, addr.clone(), portion)?;
+    // slash the vesting stake, if any
+    let vesting_slashed = vesting_stake.map(|s| s * portion).unwrap_or_default();
+    let new_vesting_stake = STAKE_VESTING.update(deps.storage, &addr, |stake| -> StdResult<_> {
+        Ok(stake.unwrap_or_default().sub(vesting_slashed))
+    })?;
```

Comment:
So again, both those updates might still be performed on 0 (default of `Uint128`)?

Reply:
Check the latest version. Here I can optimise the call, because I check on existing stake, not on proposed changes.

Reply:
No, I cannot. This is now introducing a bug. Snd now I wee why I wrote it that way.

Will revert, add tests, and properly fix this, trying to optimise saves (but not loads) in the process.

Reply:
> No, I cannot. This is now introducing a bug. ~And now I wee why I wrote it that way...~ (it was just by chance / to avoid branches)

Yes, it can, because the staked amounts are loaded at the beginning.

> 
> Will ~~revert,~~ add tests, and properly fix this; trying to optimise saves (but not loads) in the process.

Done.